datafusion/core/src/datasource/physical_plan/parquet/mod.rs

```diff
@@ -532,10 +342,648 @@ impl ParquetExec {
+///
+/// ```
+/// # use std::sync::Arc;
+/// # use arrow_schema::{Schema, SchemaRef};
+/// # use datafusion::datasource::listing::PartitionedFile;
+/// # use datafusion::datasource::data_source::FileSourceConfig;
+/// # use datafusion::datasource::physical_plan::parquet::ParquetAccessPlan;
+/// # use datafusion::datasource::physical_plan::{FileScanConfig, ParquetConfig};
+/// # use datafusion_execution::object_store::ObjectStoreUrl;
+/// # use datafusion_physical_plan::source::DataSourceExec;
+///
+/// # fn schema() -> SchemaRef {
+/// #   Arc::new(Schema::empty())
+/// # }
+/// // create an access plan to scan row group 0, 1 and 3 and skip row groups 2 and 4
+/// let mut access_plan = ParquetAccessPlan::new_all(5);
+/// access_plan.skip(2);
+/// access_plan.skip(4);
+/// // provide the plan as extension to the FileScanConfig
+/// let partitioned_file = PartitionedFile::new("my_file.parquet", 1234)
+///   .with_extensions(Arc::new(access_plan));
+/// // create a FileScanConfig to scan this file
+/// let file_scan_config = FileScanConfig::new(ObjectStoreUrl::local_filesystem(), schema())
+///     .with_file(partitioned_file);
+/// // create a ParguetConfig for file opener configurations
+/// let source_config = Arc::new(ParquetConfig::default());
+/// // this parquet DataSourceExec will not even try to read row groups 2 and 4. Additional
+/// // pruning based on predicates may also happen
+/// let exec = FileSourceConfig::new_exec(file_scan_config, source_config);
+/// ```
+///
+/// For a complete example, see the [`advanced_parquet_index` example]).
+///
+/// [`parquet_index_advanced` example]: https://github.com/apache/datafusion/blob/main/datafusion-examples/examples/advanced_parquet_index.rs
+///
+/// # Execution Overview
+///
+/// * Step 1: `DataSourceExec::execute` is called, returning a `FileStream`
+///   configured to open parquet files with a `ParquetOpener`.
+///
+/// * Step 2: When the stream is polled, the `ParquetOpener` is called to open
+///   the file.
+///
+/// * Step 3: The `ParquetOpener` gets the [`ParquetMetaData`] (file metadata)
+///   via [`ParquetFileReaderFactory`], creating a [`ParquetAccessPlan`] by
+///   applying predicates to metadata. The plan and projections are used to
+///   determine what pages must be read.
+///
+/// * Step 4: The stream begins reading data, fetching the required parquet
+///   pages incrementally decoding them, and applying any row filters (see
+///   [`Self::with_pushdown_filters`]).
+///
+/// * Step 5: As each [`RecordBatch`] is read, it may be adapted by a
+///   [`SchemaAdapter`] to match the table schema. By default missing columns are
+///   filled with nulls, but this can be customized via [`SchemaAdapterFactory`].
+///
+/// [`RecordBatch`]: arrow::record_batch::RecordBatch
+/// [`SchemaAdapter`]: crate::datasource::schema_adapter::SchemaAdapter
+/// [`ParquetMetadata`]: parquet::file::metadata::ParquetMetaData
+#[derive(Clone, Default, Debug)]
```

Comment:
if we are moving lots of code anyway, maybe this struct could go in its own module to break up the structure more